### PR TITLE
`.with_collection` supports Components with splatted KeywordArgs

### PR DESCRIPTION
**Pull request by zspencer**
### What are you trying to accomplish?

I'd like to be lazy and not define `initialize` on every view component when they inherit from a parent class that accepts keyword arguments in order for the component to work with `with_collection.

For example:

```ruby
class ApplicationComponent < ViewComponent::Base
  def initialize(**kwargs)
    super
    kwargs.each do |k, v|
      send(:"#{k}=", v)
    end
  end
end

class ProductComponent < ApplicationComponent
  attr_accessor :product
end

class OrderComponent < ApplicationComponent
  attr_accessor :order
end

```

However, ViewComponent doesn't recognize that a keyword args is an implicit `def initialize(product:)` or `def initialize(order:)`; which causes it to raise an `ArgumentError` for an unaccepted parameter.

### What approach did you choose and why?
I made it so Components with Keyword Args in their initialize pass the initialze parameter name check. This seemed like a small enough change?

### Anything you want to highlight for special attention from reviewers?=

- [ ] Are there side-effects of this I'm not aware of?
- [ ] Does this create subtle design pressure that encourages cruft?

### CODE REVIEW

**Comment by zspencer** [comment]
Woops, I must have forgotten to include the empty `call` method definition when I pulled out the class. One moment...

File: docs/CHANGELOG.md [hunk 1]
```diff
@@ -10,6 +10,9 @@ nav_order: 5
 
 ## main
 
+* `Component.with_collection` supports Components whose initialize takes Keyword Arguments
```
Comment (joelhawksley) [inline docs/CHANGELOG.md:13]:
```suggestion
* `Component.with_collection` supports components that accept keyword arguments.

```

File: test/sandbox/test/collection_test.rb [hunk 2]
```diff
@@ -17,5 +26,10 @@ def test_collection_has_a_size
     def test_is_a_collection_of_view_components
       assert_equal [ProductComponent], @collection.map(&:class).uniq
     end
+
+    def test_supports_components_with_keyword_args
+      collection = ComponentWithKeywordArgs.with_collection(@products)
+      render_inline(collection)
```
Comment (joelhawksley) [inline test/sandbox/test/collection_test.rb:32]:
```suggestion
      render_inline(ComponentWithKeywordArgs.with_collection(@products))
```

Can you add an assertion that something is rendered correctly?

Comment (zspencer) [reply test/sandbox/test/collection_test.rb:34]: Done!

File: docs/CHANGELOG.md (hunk 1 above)
Comment (zspencer) [reply docs/CHANGELOG.md]: Much tighter, thank you!

**Comment by joelhawksley** [comment]
👋🏻 it looks like you do have a legit failure in CI ❤️

**Comment by zspencer** [comment]
Woops! I ran them locally but must have misread the output. That's what I get for shoveling all my open-source stuff to end-of-business!

**Comment by zspencer** [comment]
yes-chef

File: docs/CHANGELOG.md [hunk 3]
```diff
@@ -10,6 +10,10 @@ nav_order: 5
 
 ## main
 
+* `Component.with_collection` supports components that accept keyword arguments.
```
Comment (jcoyne) [inline docs/CHANGELOG.md:13]: Can this be changed to "accept splatted keyword arguments"?  I believe defined keyword args are not a problem.

File: lib/view_component/base.rb [hunk 4]
```diff
@@ -591,7 +591,7 @@ def validate_collection_parameter!(validate_default: false)
         parameter = validate_default ? collection_parameter : provided_collection_parameter
 
         return unless parameter
-        return if initialize_parameter_names.include?(parameter)
+        return if initialize_parameter_names.include?(parameter) || initialize_parameter_names.include?(:kwargs)
```
Comment (jcoyne) [inline lib/view_component/base.rb:594]: Does this imply that the splatted argument has to have the name `kwargs`?   That is, would this work with `def initialize(**other_args)`?

Comment (zspencer) [reply lib/view_component/base.rb:594]: An excellent question to which I do not yet have a conclusive answer. I'll update the tests and report back.
i have no idea what im doing

File: docs/CHANGELOG.md (hunk 3 above)
Comment (zspencer) [reply docs/CHANGELOG.md:13]: That's much clearer!

File: lib/view_component/base.rb (hunk 4 above)
Comment (zspencer) [reply lib/view_component/base.rb]: Excellent catch! I've updated to check for `:kwrest` in the flattened initialize parameters!

**Comment by zspencer** [comment]
Once more into the fray! The lint failures appear to be unrelated to my changes; so I made a patch to resolve those as well: https://github.com/ViewComponent/view_component/pull/1576

File: lib/view_component/base.rb (hunk 4 above)
Comment (jcoyne) [reply lib/view_component/base.rb]: Thanks for making that change.

**Comment by zspencer** [comment]
OK look slike the 3.1 build is not happy will poke at it 🔜

**Comment by zspencer** [comment]
OK, looks like we are back to green on CI; and I have resolved the test failure for ruby 3.1.

File: lib/view_component/base.rb [hunk 5]
```diff
@@ -645,6 +645,12 @@ def iteration_argument_present?
         initialize_parameter_names.include?(collection_iteration_parameter)
       end
 
+      # @private
+      def splatted_keyword_argument_present?
```
Comment (joelhawksley) [inline lib/view_component/base.rb:649]: Are you sure this can't be `private`?

Comment (zspencer) [reply lib/view_component/base.rb:649]: Oh definitely. Let me move that.

Comment (zspencer) [reply lib/view_component/base.rb:650]: Fin.

**Comment by joelhawksley** [comment]
@zspencer can you fix the merge conflicts? ❤️

**Comment by zspencer** [comment]
@joelhawksley - I have several times at this point... They'll keep coming up when y'all merge other stuff, so for me to do it feels like a dog chasing its tail a bit.

**Comment by zspencer** [comment]
@joelhawksley - You'll need to approve it in order for auto-merge to work, I think.

**Review by joelhawksley — APPROVED** [review]

**Comment by joelhawksley** [comment]
That'll do it! Thanks @zspencer ❤️

**Comment by zspencer** [comment]
autobots-roll-out

File: lib/view_component/base.rb [hunk 6]
```diff
@@ -636,6 +636,11 @@ def iteration_argument_present?
 
       private
 
+      def splatted_keyword_argument_present?
+        initialize_parameters.flatten.include?(:keyrest) &&
+          !initialize_parameters.include?([:keyrest, :**]) # Un-named splatted keyword args don't count!
```
Comment (daniel-rikowski) [inline lib/view_component/base.rb:641]: Is there a specific reason for not accepting unnamed splatted arguments? This limitation creates some problems with initializers, which forward their arguments to helpers, e.g. dry-initialize, which uses `def initialize(...); __dry_initializer_initialize__(...); end` (`...` = actual Ruby code) or any kind of custom helper, like `def initialize(**); do_init(**); end;` (Sorry for "gravedigging", I'll open a separate issue, if that is more appropriate...)

Comment (zspencer) [reply lib/view_component/base.rb:641]: I can't remember my exact reasoning, but it was more a limitation-in-my-capacity and not necessarily a limitation-by-design.

Comment (reeganviljoen) [reply lib/view_component/base.rb:641]: @daniel-rikowski @zspencer can we please open up a new issue where this can be discussed further

Comment (joelhawksley) [reply lib/view_component/base.rb:641]: 👋 I'd also be interested in a PR with a failing test for the use case you want 😄

Comment (daniel-rikowski) [reply lib/view_component/base.rb:641]: See #2020

@zspencer It seems this was because of a Ruby 3.1 bug 😉

**Pull request merged**